Using Optional as a return type with Moshi converter crashes with IllegalArgumentException

What kind of issue is this?

 - [ ] Question. This issue tracker is not the place for questions. If you want to ask how to do
       something, or to understand why something isn't working the way you expect it to, use Stack
       Overflow. https://stackoverflow.com/questions/tagged/retrofit

 - [x] Bug report. If you’ve found a bug, spend the time to write a failing test. Bugs with tests
       get fixed. Here’s an example: https://gist.github.com/swankjesse/6608b4713ad80988cdc9

 - [ ] Feature Request. Start by telling us what problem you’re trying to solve. Often a solution
       already exists! Don’t send pull requests to implement new features without first getting our
       support. Sometimes we leave features out on purpose to keep the project small.

Hi, I'm trying to implement a Retrofit API with a function which returns a class wrapped in `Optional` and I'm using Moshi to parse the JSON. Here's the relevant source code for the interface:

```
interface PostmanEchoIpApi {
    @GET("/ip")
    suspend fun getIp(): Optional<PostmanResponse>
}

@JsonClass(generateAdapter = true)
data class PostmanResponse(val ip: String)
```
And here is the relevant code where's it used:
```
val retrofit = Retrofit.Builder()
    .baseUrl("https://postman-echo.com/")
    .addConverterFactory(MoshiConverterFactory.create())
    .build()
val postmanEchoIpApi = retrofit.create(PostmanEchoIpApi::class.java)
lifecycleScope.launch {
    val ip = postmanEchoIpApi.getIp()
}
```
But the app seems to be crashing with:
```
Caused by: java.lang.IllegalArgumentException: Platform class java.util.Optional in java.util.Optional<com.test.PostmanResponse> requires explicit JsonAdapter to be registered
   at com.squareup.moshi.ClassJsonAdapter$1.create(ClassJsonAdapter.java:75)
   at com.squareup.moshi.Moshi.adapter(Moshi.java:145)
   at com.squareup.moshi.Moshi.adapter(Moshi.java:105)
   at retrofit2.converter.moshi.MoshiConverterFactory.responseBodyConverter(MoshiConverterFactory.java:89)
   at retrofit2.Retrofit.nextResponseBodyConverter(Retrofit.java:362)
   at retrofit2.Retrofit.responseBodyConverter(Retrofit.java:345)
   at retrofit2.HttpServiceMethod.createResponseConverter(HttpServiceMethod.java:124)
   at retrofit2.HttpServiceMethod.parseAnnotations(HttpServiceMethod.java:85)
   at retrofit2.ServiceMethod.parseAnnotations(ServiceMethod.java:39)
   at retrofit2.Retrofit.loadServiceMethod(Retrofit.java:202)
   at retrofit2.Retrofit$1.invoke(Retrofit.java:160)
   at java.lang.reflect.Proxy.invoke(Proxy.java:1006)
   at $Proxy3.getIp(Unknown Source)
```
I think this is happening because the Moshi converter is added before the platform aware converter which handles the `Optional` type and since the Moshi converter assumes that it can handle all types, it crashes.  Shouldn't the platform aware converters be added after the built-in converters but before the rest of the converters?

I've pushed out a commit at https://github.com/tejas-n/retrofit/commit/4d3e1169f76d860a12da2262d1d112129a793e98 which seems to be fixing my issue.

Can you let me know if this is a valid issue and the commit that I pushed out is the right way to solve it, please? I'll create a PR if that's the case.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -8,7 +8,7 @@
 
 **Changed**
 
- - Nothing yet!
+ - The built-in `OptionalConverterFactory` is now public to allow installing it before other converters which consume all types (e.g., Moshi, Gson, Jackson, etc.).
 
 **Fixed**
 

--- a/retrofit/src/main/java/retrofit2/OptionalConverterFactory.java
+++ b/retrofit/src/main/java/retrofit2/OptionalConverterFactory.java
@@ -25,9 +25,24 @@ import javax.annotation.Nullable;
 import okhttp3.ResponseBody;
 import org.codehaus.mojo.animal_sniffer.IgnoreJRERequirement;
 
+/**
+ * A {@link Converter.Factory} which supports Java's {@link Optional} to wrap null values from
+ * another converter.
+ * <p>
+ * This factory is installed by default on the JVM and Android API 24+. If you are using another
+ * converter which tries to serialize all types, such as Moshi or Gson, the default installation
+ * of this factory never gets a chance to run. To work around this, you can explicitly install this
+ * factory before your serialization library converter.
+ */
 @IgnoreJRERequirement // Only added when Optional is available (Java 8+ / Android API 24+).
 @TargetApi(24)
-final class OptionalConverterFactory extends Converter.Factory {
+public final class OptionalConverterFactory extends Converter.Factory {
+  public static OptionalConverterFactory create() {
+    return new OptionalConverterFactory();
+  }
+
+  OptionalConverterFactory() {}
+
   @Override
   public @Nullable Converter<ResponseBody, ?> responseBodyConverter(
       Type type, Annotation[] annotations, Retrofit retrofit) {
@@ -43,7 +58,7 @@ final class OptionalConverterFactory extends Converter.Factory {
 
   @IgnoreJRERequirement
   static final class OptionalConverter<T> implements Converter<ResponseBody, Optional<T>> {
-    final Converter<ResponseBody, T> delegate;
+    private final Converter<ResponseBody, T> delegate;
 
     OptionalConverter(Converter<ResponseBody, T> delegate) {
       this.delegate = delegate;

